arch/rv32i/src/syscall.rs

```diff
@@ -422,45 +376,39 @@ impl kernel::syscall::UserspaceKernelBoundary for SysCall {
 
           // We have already stored the app registers in the trap handler. We
           // can restore the kernel registers before resuming kernel code.
-          lw   x1,  3*4(sp)
-          lw   x3,  4*4(sp)
-          lw   x4,  5*4(sp)
-          lw   x5,  6*4(sp)
-          lw   x6,  7*4(sp)
-          lw   x7,  8*4(sp)
-          lw   x8,  9*4(sp)
-          lw   x9,  10*4(sp)
-          lw   x10, 11*4(sp)
-          lw   x11, 12*4(sp)
-          lw   x12, 13*4(sp)
-          lw   x13, 14*4(sp)
-          lw   x14, 15*4(sp)
-          lw   x15, 16*4(sp)
-          lw   x16, 17*4(sp)
-          lw   x17, 18*4(sp)
-          lw   x18, 19*4(sp)
-          lw   x19, 20*4(sp)
-          lw   x20, 21*4(sp)
-          lw   x21, 22*4(sp)
-          lw   x22, 23*4(sp)
-          lw   x23, 24*4(sp)
-          lw   x24, 25*4(sp)
-          lw   x25, 26*4(sp)
-          lw   x26, 27*4(sp)
-          lw   x27, 28*4(sp)
-          lw   x28, 29*4(sp)
-          lw   x29, 30*4(sp)
-          lw   x30, 31*4(sp)
-          lw   x31, 32*4(sp)
-
-          addi sp, sp, 34*4   // Reset kernel stack pointer
+          //   x2            // sp -> loaded from mscratch by the trap handler
+          lw   x3,  3*4(sp)  // gp (can't be clobbered / used as an operand)
+          lw   x4,  4*4(sp)  // tp (can't be clobbered / used as an operand)
+          lw   x8,  5*4(sp)  // fp (can't be clobbered / used as an operand)
+          lw   x9,  6*4(sp)  // s1 (used internally by LLVM)
+
+          lw   a0,  1*4(sp)  // Restore the the process state pointer such that
+                             // we don't need to mark it as clobbered.
+                             // Otherwise, this would cause Rust to stack a
+                             // register which we already manually save.
+
+          addi sp, sp, 8*4   // Reset kernel stack pointer
           ",
 
           // The register to put the state struct pointer in is not
           // particularly relevant, however we must avoid using t0
           // as that is overwritten prior to being accessed
           // (although stored and later restored) in the assembly
           in("a0") state as *mut Riscv32iStoredState,
+
+          // Clobber all registers which can be marked as clobbered, except
+          // for `a0` / `x10`. By making it retain the value of `&mut state`,
+          // which we need to stack manually anyway, we can avoid Rust/LLVM
+          // stacking it redundantly for us.
+          lateout("x1") _, lateout("x5") _, lateout("x6") _,
```

Comment:
Could you explain the purpose of using `lateout` here instead of `out`?

It seems like the only difference is that `lateout` registers could be reused input registers, after they've already be read and utilized in the code and their value no longer matters. That doesn't seem to apply here though, since all registers are explicitly named, and there's only one input register anyways, which doesn't overlap with any of the `lateout` registers.

Reply:
Okay, that makes sense to me then. I thought there was some deeper reason why they _needed_ to be `lateout`.

Reply:
The Rust docs say that lateout "can be used on any output that is written only after all inputs have been consumed.", which applies to us here. So it does seem like the one to use

Reply:
> "can be used on any output that is written only after all inputs have been consumed.", which applies to us here

While technically true, I don't think this is what the Rust developers intended it to mean. I read it as simply a notice that you'd need to be careful with `lateout` to not clobber a register after you've read its contents. There's really no reason to use `lateout` if a register is solely defined as `in` and it won't be clobbered by the assembly. For cases where a register is solely defined as `out`, Rust will not rely on its contents not changing, and allow subsequent code to access this register's value. However, it won't work when a register is also defined as `in`; the inability to use `in` + `out` on a single register is the sole purpose of `lateout` (which we don't do here).

Reply:
It seems a little unclear to me which is preferable (I believe either would be "correct"). On one hand, using `out` gives the benefit leon mentioned -- you get a warning if you later change the input registers such that there is an alias, and that is nice because you probably do not want to unintentionally alias your in and output registers. But using `lateout` presumably decreases register pressure a little, since the compiler can feel free to keep whatever was already in the `lateout` register up until the final input is read.

That said, I am not sure the compiler would even make use of that second optimization -- it seems possible that `lateout` is only beneficial if you are not selecting a specific register, and instead using the generic `reg`, such that the compiler can just reuse any of the input registers.

Reply:
> it seems possible that lateout is only beneficial if you are not selecting a specific register, and instead using the generic reg, such that the compiler can just reuse any of the input registers.

After doing some reading, this was my takeaway. I think it's irrelevant when picking specific registers.